MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/BuyClientError.java

```diff
@@ -0,0 +1,182 @@
+/*
+ *   The MIT License (MIT)
+ *
+ *   Copyright (c) 2015 Shopify Inc.
+ *
+ *   Permission is hereby granted, free of charge, to any person obtaining a copy
+ *   of this software and associated documentation files (the "Software"), to deal
+ *   in the Software without restriction, including without limitation the rights
+ *   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *   copies of the Software, and to permit persons to whom the Software is
+ *   furnished to do so, subject to the following conditions:
+ *
+ *   The above copyright notice and this permission notice shall be included in
+ *   all copies or substantial portions of the Software.
+ *
+ *   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *   THE SOFTWARE.
+ */
+package com.shopify.buy.dataprovider;
+
+import android.text.TextUtils;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import retrofit2.Response;
+
+/**
+ * Exception thrown when API call failed, and be returned in {@link Callback#failure(BuyClientError)}
+ */
+public final class BuyClientError extends RuntimeException {
+
+    /**
+     * Represents any errors returning from the backend
+     */
+    public static final int ERROR_TYPE_API = -2;
+
+    /**
+     * Represents any network raletd errors
+     */
+    public static final int ERROR_TYPE_NETWORK = -1;
+
+    /**
+     * Represent any unknown errors
+     */
+    public static final int ERROR_TYPE_UNKNOWN = 0;
+
+    private final int type;
+
+    private final Response retrofitResponse;
+
+    private final JSONObject errorsRootJsonObject;
+
+    public BuyClientError(final Response retrofitResponse, final Throwable throwable) {
```

Comment:
I don't see anywhere that we set the response and the throwable.  I would prefer to see two constructors